test/java/seedu/addressbook/commands/ViewCommandTest.java

```diff
@@ -0,0 +1,161 @@
+
+        TypicalPersons td = new TypicalPersons();
+        typicalAddressBook = td.getTypicalAddressBook();
+        displayListWithAllTypicalPersons = Arrays.asList(td.getTypicalPersons());
+        listWithSome = Arrays.asList(td.amy, td.candy, td.dan);
+    }
+
+    @Test
+    public void execute_invalidIndex_returnsInvalidIndexMessage() {
+        // empty addressbook
+        assertViewErrorInvalidIndex(emptyAddressBook, emptyDisplayList, 1);
+
+        // non-empty addressbook
+        assertViewErrorInvalidIndex(typicalAddressBook, displayListWithAllTypicalPersons, -1);
+        assertViewErrorInvalidIndex(typicalAddressBook, displayListWithAllTypicalPersons, 0);
+        assertViewErrorInvalidIndex(typicalAddressBook, displayListWithAllTypicalPersons, displayListWithAllTypicalPersons.size() + 1);
+    }
+
+    @Test
+    public void execute_personNotInAddressBook_returnsPersonNotInAddressBookMessage() throws Exception {
+        // generate person not in addressbook, add to displayList
+        ReadOnlyPerson someone = new Person(new Name("me"),
+                                            new Phone("123", true),
+                                            new Email("some@hey.go", true),
+                                            new Address("nus", false),
+                                            new UniqueTagList(Collections.emptySet()));
+        List <ReadOnlyPerson> displayListWithExtraPerson = new ArrayList<ReadOnlyPerson>(displayListWithAllTypicalPersons);
+        displayListWithExtraPerson.add(someone);
+
+        // empty addressbook
+        assertViewErrorPersonNotInAddressBook(emptyAddressBook, displayListWithExtraPerson, 1);
+
+        // non-empty addressbook
+        assertViewErrorPersonNotInAddressBook(typicalAddressBook, displayListWithExtraPerson, displayListWithExtraPerson.size());
+    }
+
+    @Test
+    public void execute_validIndex_returnsPersonDetails() {
+        // person with no private information
+        assertViewSuccess(typicalAddressBook, displayListWithAllTypicalPersons, 1);
+
+        // person with some private information
+        assertViewSuccess(typicalAddressBook, displayListWithAllTypicalPersons, 2);
+
+        // person with all private information
+        assertViewSuccess(typicalAddressBook, displayListWithAllTypicalPersons, 4);
+
+        // addressbook has more people than displayList
+        // This can happen when view from filtered list caused by some commands(eg. FindCommand)
+        assertViewSuccess(typicalAddressBook, listWithSome, 1);
+    }
+
+    private Command generateViewCommand(AddressBook addressBook, List<ReadOnlyPerson> relevantPersons, int targetIndex) {
+        Command command = new ViewCommand(targetIndex);
+        command.setData(addressBook, relevantPersons);
+
+        return command;
+    }
+
+    private Command generateViewAllCommand(AddressBook addressBook, List<ReadOnlyPerson> relevantPersons, int targetIndex) {
```

Comment:
The code of `generateViewCommand()` and `generateViewAllCommand()` look very similar, which hints that they can be combined into a more generic method.